File: python/0039.py
class Solution:
    def __init__(self):
        self.ans = []

    def go(self, candidates, ind, target, now):
        if ind == len(candidates):
            if target == 0:
                self.ans.append(now[:])
            return
        for i in range(target // candidates[ind] + 1):
            self.go(candidates, ind + 1, target - i * candidates[ind], now + [candidates[ind]] * i)

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.go(candidates, 0, target, [])
        return self.ans


if __name__ == '__main__':
    a = []
    print(a + [1] * 0)

File: rust/src/bin/0039.rs
struct Solution;

fn go(cand: &Vec<i32>, target: i32, ind: usize, s: i32, ans: &mut Vec<Vec<i32>>, now: &mut Vec<i32>) {
    if s > target {
        return;
    }
    if ind == cand.len() {
        if s == target {
            ans.push(now.clone());
        }
        return;
    }
    let i = cand[ind];
    let mut cnt = 0;
    // println!("ind={} i={} s={} now={:?}", ind, i, s, now);
    while s + cnt * i <= target {
        go(cand, target, ind + 1, s + cnt * i, ans, now);
        now.push(i);
        cnt += 1;
    }
    while cnt > 0 {
        now.pop();
        cnt -= 1;
    }
}

impl Solution {
    pub fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        candidates.sort();
        candidates.dedup_by_key(|x| { *x });
        let mut ans: Vec<Vec<i32>> = vec!();
        go(&candidates, target, 0, 0, &mut ans, &mut vec!());
        ans
    }
}

fn main() {
    let ans = Solution::combination_sum(vec!(2, 3, 6, 7), 7);
    println!("{:?}", ans);
}